File: mobile/ios/OpenU60/Features/RouterSettings/Telemetry/TelemetryBlockerViewModel.swift
import SwiftUI

@Observable
@MainActor
final class TelemetryBlockerViewModel {
    var filterConfig: DomainFilterConfig = .empty
    var isLoading: Bool = false
    var message: String?
    var messageIsError: Bool = false
    var newDomain: String = ""

    private let client: AgentClient
    private let authManager: AuthManager

    init(client: AgentClient, authManager: AuthManager) {
        self.client = client
        self.authManager = authManager
    }

    func refresh() async {
        isLoading = true
        message = nil

        do {
            let data = try await client.getJSON("/api/router/domain-filter")
            filterConfig = TelemetryParser.parseDomainFilter(data)
        } catch {
            showMessage("Failed to load filters: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func toggleFilter(enabled: Bool) async {
        isLoading = true

        do {
            let _ = try await client.putJSON("/api/router/domain-filter", body: ["enable": enabled ? "1" : "0"])
            showMessage("Domain filter \(enabled ? "enabled" : "disabled")", isError: false)
            filterConfig.enabled = enabled
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func addDomain(_ domain: String) async {
        let trimmed = domain.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            showMessage("Enter a domain name", isError: true)
            return
        }

        isLoading = true

        do {
            let _ = try await client.putJSON("/api/router/domain-filter", body: [
                "action": "add",
                "domain": trimmed,
                "enabled": "1"
            ])
            newDomain = ""
            showMessage("Added \(trimmed)", isError: false)
            filterConfig.rules.append(DomainFilterRule(id: UUID().uuidString, domain: trimmed, enabled: true))
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func removeDomain(_ rule: DomainFilterRule) async {
        isLoading = true

        do {
            let _ = try await client.putJSON("/api/router/domain-filter", body: [
                "action": "delete",
                "id": rule.id
            ])
            showMessage("Removed \(rule.domain)", isError: false)
            filterConfig.rules.removeAll { $0.id == rule.id }
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func blockAllTelemetry() async {
        isLoading = true
        let existingDomains = Set(filterConfig.rules.map(\.domain))

        var added = 0
        for domain in TelemetryParser.knownTelemetryDomains {
            guard !existingDomains.contains(domain) else { continue }
            do {
                let _ = try await client.putJSON("/api/router/domain-filter", body: [
                    "action": "add",
                    "domain": domain,
                    "enabled": "1"
                ])
                added += 1
                filterConfig.rules.append(DomainFilterRule(id: UUID().uuidString, domain: domain, enabled: true))
            } catch {
                // Continue with remaining domains
            }
        }

        if added > 0 {
            showMessage("Blocked \(added) telemetry domain\(added == 1 ? "" : "s")", isError: false)
        } else {
            showMessage("All telemetry domains already blocked", isError: false)
        }

        isLoading = false
    }

    private func showMessage(_ text: String, isError: Bool) {
        message = text
        messageIsError = isError
    }
}

File: mobile/android/OpenU60/app/src/main/java/com/openu60/feature/router/telemetry/TelemetryBlockerViewModel.kt
package com.openu60.feature.router.telemetry

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.openu60.core.model.DomainFilterConfig
import com.openu60.core.model.TelemetryParser
import com.openu60.core.network.AgentClient
import com.openu60.core.network.AgentError
import com.openu60.core.network.AuthManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class TelemetryBlockerState(
    val config: DomainFilterConfig = DomainFilterConfig.empty,
    val isLoading: Boolean = false,
    val message: String? = null,
    val messageIsError: Boolean = false,
    val newDomain: String = "",
)

@HiltViewModel
class TelemetryBlockerViewModel @Inject constructor(
    private val agentClient: AgentClient,
    private val authManager: AuthManager,
) : ViewModel() {

    private val _state = MutableStateFlow(TelemetryBlockerState())
    val state: StateFlow<TelemetryBlockerState> = _state.asStateFlow()

    fun refresh() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val data = agentClient.getJSON("/api/firewall/domain-filter")
                val config = TelemetryParser.parseDomainFilter(data)
                _state.value = _state.value.copy(config = config, isLoading = false)
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) refresh() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun toggleFilter(enabled: Boolean) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.putJSON("/api/firewall/domain-filter", mapOf("enable" to if (enabled) "1" else "0"))
                _state.value = _state.value.copy(message = "Filter ${if (enabled) "enabled" else "disabled"}", messageIsError = false)
                refresh()
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) toggleFilter(enabled) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun updateNewDomain(value: String) {
        _state.value = _state.value.copy(newDomain = value)
    }

    fun addRule(domain: String) {
        if (domain.isBlank()) return
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.postJSON("/api/firewall/domain-filter/rule", mapOf("domain" to domain))
                _state.value = _state.value.copy(newDomain = "", message = "Rule added", messageIsError = false)
                refresh()
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) addRule(domain) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun removeRule(id: String) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.deleteJSON("/api/firewall/domain-filter/rule", mapOf("id" to id))
                _state.value = _state.value.copy(message = "Rule removed", messageIsError = false)
                refresh()
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) removeRule(id) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun blockAllTelemetry() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val existing = _state.value.config.rules.map { it.domain }.toSet()
                var added = 0
                for (domain in TelemetryParser.knownTelemetryDomains) {
                    if (domain !in existing) {
                        agentClient.postJSON("/api/firewall/domain-filter/rule", mapOf("domain" to domain))
                        added++
                    }
                }
                _state.value = _state.value.copy(
                    message = if (added > 0) "Blocked $added telemetry domains" else "All telemetry domains already blocked",
                    messageIsError = false,
                )
                refresh()
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) blockAllTelemetry() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    private fun setError(msg: String?) {
        _state.value = _state.value.copy(isLoading = false, message = msg ?: "Unknown error", messageIsError = true)
    }
}
